### PR TITLE
Update to default English GenerativeQA template

### PR DESCRIPTION
**Pull request by sjrl**
This takes input from the Sol-team on what we have learned to perform well for Generative QA. 

Some changes:
- Remove the AnswerGenerator node in favor of the PromptNode
- Use a better default prompt to provide to the model
- Use gpt-3.5-turbo over text-davinci-003 for speed and reduced cost
- Use a Hybrid Retrieval with a Reranker. This has provided the best recall at 3 results so far in practice which is essential since we find providing too many documents to the Generative model can degrade performance

### CODE REVIEW

**Comment by tholor** [comment]
Good that we set temperature=0 now! works well better in my experience for most of our use cases....

File: pipelines/GenerativeQuestionAnswering_gpt.yaml [hunk 1]
```diff
@@ -2,32 +2,63 @@
 # This is a friendly editor that helps you create your pipelines with autosuggestions. To use them, press control + space on your keyboard.
 # Whenever you need to specify a model, this editor helps you out as well. Just type your Hugging Face organization and a forward slash (/) to see available models.
 
-# This is a Generative Question Answering pipeline for English with a good vector-based Retriever and OpenAI's GPT-3 model as a Generator
+# This is a Generative Question Answering pipeline for English with a good vector-based Retriever and OpenAI's GPT-3.5 model as a PromptNode
 version: '1.17.1'
-name: 'GenerativeQuestionAnswering_GPT-3'
+name: 'GenerativeQuestionAnswering_GPT-35'
```
Comment (agnieszka-m) [inline pipelines/GenerativeQuestionAnswering_gpt.yaml:7]:
```suggestion
name: 'GenerativeQuestionAnswering_GPT-3.5'
```

File: pipelines/GenerativeQuestionAnswering_gpt.yaml [hunk 2]
```diff
@@ -2,32 +2,63 @@
...
+  - name: Reranker # Uses a cross-encoder model to rerank the documents returned by the two retrievers
+    type: SentenceTransformersRanker
+    params:
+      model_name_or_path: cross-encoder/ms-marco-MiniLM-L-12-v2 # Best performing model on the cross-encoders website https://www.sbert.net/docs/pretrained-models/ce-msmarco.html
       top_k: 3 # The number of results to return
+      batch_size: 20  # Try to keep this number equal or larger to the sum of the top_k of the two retrievers so all docs are processed at once
+  - name: qa_template
+    type: PromptTemplate
+    params:
+      name: qa-new
```
Comment (agnieszka-m) [inline pipelines/GenerativeQuestionAnswering_gpt.yaml:40]: can we call it something like: question-answering-hybrid or generative-question-answering? (users are going to see this name)

File: pipelines/GenerativeQuestionAnswering_gpt.yaml [hunk 3]
```diff
@@ -2,32 +2,63 @@
...
+        \ Reference every article that you find useful for your answer. Make sure your reference number corresponds to the index of the source document.\ 
+        \ If you cannot answer the question based on the document, you say: 'The answer is not in the document.'\
+        \ Question: {query}\
+        \ Answer:\
+        \  \n"
+  - name: PromptNode
+    type: PromptNode
+    params:
+      default_prompt_template: qa_template
+      max_length: 400
```
Comment (agnieszka-m) [inline pipelines/GenerativeQuestionAnswering_gpt.yaml:58]:
```suggestion
      max_length: 400 # The maximum number of tokens the generated answer can have
```

File: pipelines/GenerativeQuestionAnswering_gpt.yaml [hunk 4]
```diff
@@ -2,32 +2,63 @@
...
+      model_kwargs:
```
Comment (agnieszka-m) [inline pipelines/GenerativeQuestionAnswering_gpt.yaml:59]:
```suggestion
      model_kwargs: # Specifies additional model settings
```

**Comment by sjrl** [comment]
Yes definitely! We've been using temperature=0 for most of our client pipelines as well (e.g. Manz and Zeit-Online)

File: pipelines/GenerativeQuestionAnswering_gpt.yaml [hunk 5]
```diff
@@ -2,32 +2,63 @@
...
+        temperature: 0
```
Comment (agnieszka-m) [inline pipelines/GenerativeQuestionAnswering_gpt.yaml:60]:
```suggestion
        temperature: 0 # Lower temperature works best for fact-based qa
```

**Comment by sjrl** [comment]
Thanks @ju-gu! I'll add your suggestions for the prompt update. 

What would you suggest a good top_k to be then? Of course we will be limited by the token limit and the number of docs, but for the current preprocessor shown in the pipeline (splitting at 250 words) do you have an idea of what a sensible top_k would be?

File: pipelines/GenerativeQuestionAnswering_gpt.yaml (hunk 1 above)
Comment (sjrl) [reply pipelines/GenerativeQuestionAnswering_gpt.yaml:7]: Just double checking, we can support a period symbol (`.`) in the name of a pipeline in dC, right? I'm afraid that might break something. @tholor @erendabanlioglu do you happen to know?

File: pipelines/GenerativeQuestionAnswering_gpt.yaml (hunk 2 above)
Comment (sjrl) [reply pipelines/GenerativeQuestionAnswering_gpt.yaml:40]: I'd be happy to change to `generative-question-answering`

**Comment by mathislucka** [comment]
Also consider costs when bumping up the top_k

**Comment by ju-gu** [comment]
> Also consider costs when bumping up the top_k

I think costs are not so much a factor if we use turbo 3.5.

File: pipelines/GenerativeQuestionAnswering_gpt.yaml (hunk 1 above)
Comment (sjrl) [reply pipelines/GenerativeQuestionAnswering_gpt.yaml]: I checked on dC and the period symbol appears to work just fine.

Comment (agnieszka-m) [reply pipelines/GenerativeQuestionAnswering_gpt.yaml]: yup, checked that too before making this comment, didn't break anything :)

Comment (sjrl) [reply pipelines/GenerativeQuestionAnswering_gpt.yaml]: Thanks! :) I made the change in a separate commit.

File: pipelines/GenerativeQuestionAnswering_gpt.yaml [hunk 6]
```diff
@@ -2,32 +2,69 @@
...
+  - name: Reranker # Uses a cross-encoder model to rerank the documents returned by the two retrievers
+    type: SentenceTransformersRanker
     params:
-      model: text-davinci-003
-      max_tokens: 200 # The maximum number of tokens allowed for each generated Answer.
-      temperature: 0.8 # Determines the randomness of the model. Higher values mean the model will take more risks
-      frequency_penalty: 0.1 # Positive values penalize new tokens based on their existing frequency in the text so far.
-      presence_penalty: 0.1 #  Positive values penalize new tokens based on whether they have already appeared in the text.
-      top_k: 3 # The number of results to return
+      model_name_or_path: cross-encoder/ms-marco-MiniLM-L-12-v2 # Best performing model on the cross-encoders website https://www.sbert.net/docs/pretrained-models/ce-msmarco.html
```
Comment (mathislucka) [inline pipelines/GenerativeQuestionAnswering_gpt.yaml:34]: Why is the doc search template using L6 but this here is using L12?

File: pipelines/GenerativeQuestionAnswering_gpt.yaml [hunk 7]
```diff
@@ -2,32 +2,69 @@
...
+      top_k: 4 # The number of results to return
+      batch_size: 20  # Try to keep this number equal or larger to the sum of the top_k of the two retrievers so all docs are processed at once
+  - name: qa_template
+    type: PromptTemplate
+    params:
+      output_parser:
+        type: AnswerParser
+      prompt: "You are a technical expert.\
```
Comment (mathislucka) [inline pipelines/GenerativeQuestionAnswering_gpt.yaml:42]: How did you test this template?

**Review by mathislucka — APPROVED** [review]
nothing blocking, more information on how you tested the template would be nice

File: pipelines/GenerativeQuestionAnswering_gpt.yaml (hunk 6 above)
Comment (sjrl) [reply pipelines/GenerativeQuestionAnswering_gpt.yaml:34]: Good catch, I found on the ST website that L6 is as accurate as L12 (on MS-Marco) but is twice as fast, so I think we should use L6 in our templates.

File: pipelines/GenerativeQuestionAnswering_gpt.yaml (hunk 7 above)
Comment (sjrl) [reply pipelines/GenerativeQuestionAnswering_gpt.yaml:42]: This was a template I picked up from the general-demos that I believe @bglearning developed and was then supplemented with the suggestions from @ju-gu which are a close to direct translation from the prompt used for our legal client.

**Comment by sjrl** [comment]
Will wait to merge once 1.18 is in dC. This template is already updated to use the new variable names for the PromptTemplate

**Review by erendabanlioglu — COMMENTED** [review]
Should this be a new template? I remember we wanted to have one with answer gen node and one with prompt node. @mathislucka

**Comment by sjrl** [comment]
I believe the plan is to deprecate the AnswerGen node and the AnswerGen node does not work with the prompt explorer.

**Comment by mathislucka** [comment]
> Should this be a new template? I remember we wanted to have one with answer gen node and one with prompt node. @mathislucka

Yes, answer gen should go away now as it's not supported by our prompt explorer.

File: pipelines/GenerativeQuestionAnswering_gpt.yaml [hunk 8]
```diff
@@ -2,32 +2,70 @@
 # This is a friendly editor that helps you create your pipelines with autosuggestions. To use them, press control + space on your keyboard.
 # Whenever you need to specify a model, this editor helps you out as well. Just type your Hugging Face organization and a forward slash (/) to see available models.
 
-# This is a Generative Question Answering pipeline for English with a good vector-based Retriever and OpenAI's GPT-3 model as a Generator
-version: '1.17.1'
-name: 'GenerativeQuestionAnswering_GPT-3'
+# This is a Generative Question Answering pipeline for English with a good vector-based Retriever and OpenAI's GPT-3.5 model as a PromptNode
+version: '1.18.1'
```
Comment (erendabanlioglu) [inline pipelines/GenerativeQuestionAnswering_gpt.yaml:6]:
```suggestion
version: '1.18.0'
```

**Review by erendabanlioglu — APPROVED** [review]

**Pull request merged**